### PR TITLE
Linux

### LINKED ISSUE #484: Linux
This library should run on all of Swift’s officially supported platforms, including Linux. Linux support should be practical because this library has no real UI dependency, other than a few conveniences around default pixel densities that are easy to stub out.

Linux support will enable the command line utility in #451 to be run in the same development environment that Valhalla developers use, making it more likely that any regressions introduced in the Mapbox Directions API will be quickly caught by the server-side teams before production deployment.

#366 attempted to add Linux and SPM support and rename the library from MapboxDirections.swift to MapboxDirections all at the same time, but it got bogged down by the need to support Objective-C. Since then, we added SPM support in #362, removed Objective-C compatibility in #382, and renamed the library in #400. #366 may be of interest for a few build-related files, but most of its changes are no longer necessary. The one remaining blocker is that Polyline doesn’t support Linux yet: raphaelmor/Polyline#55.

* [x] Refresh raphaelmor/Polyline#55 and get it merged and released
* [x] Carry over relevant portions of #366:
   * .circleci/config.yml
   * .gitignore
   * Package.swift
   * Tests/MapboxDirectionsTests/XCTestManifests.swift

/cc @mapbox/navigation-ios @frederoni

### PR DESCRIPTION
**Pull request by 1ec5**
Added support for building and running this library on Linux.

Fixed compiler errors and failures in unit tests in the SPM-managed package (as an alternative to the Carthage-managed Xcode project). Increased the minimum required Swift Package Manager version from 5.0 to 5.3 so that SPM can pull in the test fixtures.

More to come:

* [x] Get SPM package to build without errors in Xcode 12 on Apple platforms
* [x] Get SPM package to build without errors on Linux
* [x] Get SPM package to pass unit tests on Linux
* [x] Upgrade to Turf v1.1.0 once it’s out
* [x] Upgrade to Polyline v5.0.2 once it’s out
* [x] Update documentation to acknowledge Linux support

Fixes #484. Reprise of #366. Depends on #487, mapbox/turf-swift#120, and mapbox/turf-swift#121.

/cc @mapbox/navigation-ios @frederoni

### CODE REVIEW

File: Sources/MapboxDirections/Extensions/CoreLocation.swift [hunk 1]
```diff
@@ -1,5 +1,9 @@
 import Foundation
+#if canImport(CoreLocation)
 import CoreLocation
+#else
+import Turf
+#endif
```
Comment (1ec5) [inline Sources/MapboxDirections/Extensions/CoreLocation.swift:6]: This PR takes advantage of the compatibility shims for Core Location types that Turf provides as of mapbox/turf-swift#30. mapbox/turf-swift#119 tracks making Turf’s Core Location compatibility shim consistent with Polyline’s so it’s less magical and more obvious how to choose a particular shim.

**Comment by 1ec5** [comment]
The Core Location compatibility shim in Turf is incomplete, triggering compiler errors on Linux:

```
[25/68] Compiling MapboxDirections CoreLocation.swift/root/project/Sources/MapboxDirections/Extensions/CoreLocation.swift:14:11: error: cannot find type 'CLLocation' in scope
extension CLLocation {
          ^~~~~~~~~~
/root/project/Sources/MapboxDirections/Extensions/CoreLocation.swift:10:19: error: 'longitude' is inaccessible due to 'internal' protection level
        return "\(longitude.rounded(to: 1e6)),\(latitude.rounded(to: 1e6))"
                  ^
/root/project/.build/checkouts/turf-swift/Sources/Turf/CoreLocation.swift:5:9: note: 'longitude' declared here
    let longitude: Double
        ^
/root/project/Sources/MapboxDirections/Extensions/CoreLocation.swift:10:49: error: 'latitude' is inaccessible due to 'internal' protection level
        return "\(longitude.rounded(to: 1e6)),\(latitude.rounded(to: 1e6))"
                                                ^
/root/project/.build/checkouts/turf-swift/Sources/Turf/CoreLocation.swift:4:9: note: 'latitude' declared here
    let latitude: Double
        ^
[26/68] Compiling MapboxDirections Double.swift/root/project/Sources/MapboxDirections/Extensions/CoreLocation.swift:14:11: error: cannot find type 'CLLocation' in scope
extension CLLocation {
          ^~~~~~~~~~
/root/project/Sources/MapboxDirections/Extensions/CoreLocation.swift:10:19: error: 'longitude' is inaccessible due to 'internal' protection level
        return "\(longitude.rounded(to: 1e6)),\(latitude.rounded(to: 1e6))"
                  ^
/root/project/.build/checkouts/turf-swift/Sources/Turf/CoreLocation.swift:5:9: note: 'longitude' declared here
    let longitude: Double
        ^
/root/project/Sources/MapboxDirections/Extensions/CoreLocation.swift:10:49: error: 'latitude' is inaccessible due to 'internal' protection level
        return "\(longitude.rounded(to: 1e6)),\(latitude.rounded(to: 1e6))"
                                                ^
/root/project/.build/checkouts/turf-swift/Sources/Turf/CoreLocation.swift:4:9: note: 'latitude' declared here
    let latitude: Double
        ^
```

mapbox/turf-swift#120 would fix these errors, but it’ll require a new release.

**Comment by 1ec5** [comment]
Every now and then, the Linux CI build fails with an odd error like:

```
[3/22] Merging module PolylineBox.swift
```

As if it’s merging together any module whose name happens to start with “Polyline”.

**Comment by 1ec5** [comment]
The failures are consistent now. If I `ssh` into the build machine and rerun `swift build`, it shows some of the same error messages as before:

```
swift build
/root/project/Sources/MapboxDirections/Extensions/Codable.swift:13:34: error: cannot convert value of type '[CLLocationCoordinate2D]' to expected argument type '[LocationCoordinate2D]'
        return encodeCoordinates(coordinates, precision: precision)
                                 ^
/root/project/Sources/MapboxDirections/Extensions/Codable.swift:13:34: note: arguments to generic parameter 'Element' ('CLLocationCoordinate2D' and 'LocationCoordinate2D') are expected to be equal
        return encodeCoordinates(coordinates, precision: precision)
                                 ^
/root/project/Sources/MapboxDirections/Extensions/Codable.swift:13:34: error: cannot convert value of type '[CLLocationCoordinate2D]' to expected argument type '[LocationCoordinate2D]'
        return encodeCoordinates(coordinates, precision: precision)
                                 ^
/root/project/Sources/MapboxDirections/Extensions/Codable.swift:13:34: note: arguments to generic parameter 'Element' ('CLLocationCoordinate2D' and 'LocationCoordinate2D') are expected to be equal
        return encodeCoordinates(coordinates, precision: precision)
                                 ^
/root/project/Sources/MapboxDirections/Extensions/GeoJSON.swift:25:87: error: type of expression is ambiguous without more context
        guard var coordinates = decodePolyline(encodedPolyline, precision: precision) as [CLLocationCoordinate2D]? else {
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/project/Sources/MapboxDirections/Extensions/GeoJSON.swift:25:87: error: type of expression is ambiguous without more context
        guard var coordinates = decodePolyline(encodedPolyline, precision: precision) as [CLLocationCoordinate2D]? else {
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
[65/65] Compiling MapboxDirections RouteStep.swift
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -37,6 +37,13 @@ commands:
                 command: bash <(curl -s https://codecov.io/bash)
 
 jobs:
+  spm-linux-job:
+    docker:
+      - image: swift:5.3
+    steps:
+      - checkout
+      - run: swift build
```
Comment (frederoni) [inline .circleci/config.yml:45]: Would be nice if we could run `swift test` on Linux but there's a few of them that require CoreLocation and other frameworks that are not available on Linux.

**Review by frederoni — APPROVED** [review]

**Comment by 1ec5** [comment]
GeoJSONTests is [crashing](https://app.circleci.com/pipelines/github/mapbox/mapbox-directions-swift/804/workflows/802f129d-5eba-45a3-b705-039d33b4d400/jobs/2371) on the Linux build bot:

```
Test Suite 'CredentialsTests' passed at 2020-11-06 16:34:18.129
	 Executed 1 test, with 0 failures (0 unexpected) in 0.0 (0.0) seconds
…
Test Suite 'DirectionsErrorTests' passed at 2020-11-06 16:34:18.130
	 Executed 3 tests, with 0 failures (0 unexpected) in 0.001 (0.001) seconds
…
Test Suite 'GeoJSONTests' started at 2020-11-06 16:32:34.314
…
Exited with signal code 4
```

**Comment by 1ec5** [comment]
```
$ swift test --filter GeoJSONTests
Test Suite 'Selected tests' started at 2020-11-06 16:42:26.061
Test Suite 'GeoJSONTests' started at 2020-11-06 16:42:26.062
Test Case 'GeoJSONTests.testInitialization' started at 2020-11-06 16:42:26.062
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:7: error: GeoJSONTests.testInitialization : XCTAssertThrowsError failed: did not throw error - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:12: error: GeoJSONTests.testInitialization : XCTAssertEqual failed: ("Optional(5)") is not equal to ("Optional(2)") - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:13: error: GeoJSONTests.testInitialization : XCTAssertEqual failed: ("-9e-05") is not equal to ("39.27665") +/- ("1e-05") - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:14: error: GeoJSONTests.testInitialization : XCTAssertEqual failed: ("1e-05") is not equal to ("-84.411389") +/- ("1e-05") - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:15: error: GeoJSONTests.testInitialization : XCTAssertEqual failed: ("-0.00382") is not equal to ("39.276635") +/- ("1e-05") - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:16: error: GeoJSONTests.testInitialization : XCTAssertEqual failed: ("-0.00224") is not equal to ("-84.411148") +/- ("1e-05") - 
Test Case 'GeoJSONTests.testInitialization' failed (0.0 seconds)
Test Case 'GeoJSONTests.testZeroLengthWorkaround' started at 2020-11-06 16:42:26.062
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:26: error: GeoJSONTests.testZeroLengthWorkaround : XCTAssertEqual failed: ("Optional(5)") is not equal to ("Optional(2)") - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:27: error: GeoJSONTests.testZeroLengthWorkaround : XCTAssertEqual failed: ("Optional(Turf.CLLocationCoordinate2D(latitude: -9e-06, longitude: 1e-06))") is not equal to ("Optional(Turf.CLLocationCoordinate2D(latitude: -0.000381, longitude: 0.119742))") - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:28: error: GeoJSONTests.testZeroLengthWorkaround : XCTAssertEqual failed: ("Optional("PANStU@@gzhF??")") is not equal to ("Optional("s{byuAnigzhF??")") - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:31: error: GeoJSONTests.testZeroLengthWorkaround : XCTAssertNoThrow failed: threw error "cannotDecodePolyline(precision: 1000000.0)" - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:33: error: GeoJSONTests.testZeroLengthWorkaround : XCTAssertEqual failed: ("Optional(5)") is not equal to ("Optional(2)") - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:34: error: GeoJSONTests.testZeroLengthWorkaround : XCTAssertEqual failed: ("Optional(Turf.CLLocationCoordinate2D(latitude: -9e-06, longitude: 1e-06))") is not equal to ("Optional(Turf.CLLocationCoordinate2D(latitude: -0.000381, longitude: 0.119742))") - 
/root/project/Tests/MapboxDirectionsTests/GeoJSONTests.swift:35: error: GeoJSONTests.testZeroLengthWorkaround : XCTAssertEqual failed: ("Optional("PANStU@@gzhF??")") is not equal to ("Optional("s{byuArigzhF??")") - 
Test Case 'GeoJSONTests.testZeroLengthWorkaround' failed (0.0 seconds)
Test Suite 'GeoJSONTests' failed at 2020-11-06 16:42:26.063
	 Executed 2 tests, with 13 failures (0 unexpected) in 0.001 (0.001) seconds
Test Suite 'Selected tests' failed at 2020-11-06 16:42:26.063
	 Executed 2 tests, with 13 failures (0 unexpected) in 0.001 (0.001) seconds
```

**Comment by 1ec5** [comment]
The culprit for the crash seems to be that `LineString(encodedPolyline:precision:)` can throw a GeometryError, which is a LocalizableError. I haven’t seen a crash so far if I turn it into a standard Error.

File: .circleci/config.yml (hunk 2 above)
Comment (1ec5) [reply .circleci/config.yml:45]: I added a `swift test` command. The tests were almost working, but they exposed a memory management bug in Polyline: raphaelmor/Polyline#71.

**Pull request merged**